### ISSUE TITLE
date pipe formatter doesn't format minutes and seconds properly in IE 11.0.9600.18350 and Edge

### ISSUE THREAD

**Issue by rickjz** [angular/angular#9524]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When using the date pipe In IE 11.0.9600.18350 with minutes/seconds it shows the whole date instead of the minutes and/or seconds.

Examples:
Date recieved from service | used binding                        | result
"2016-06-21T10:47:40Z"    | {{mydate | date:'HH'}}          | ‎12‎:‎00
"2016-06-21T10:47:40Z"    | {{mydate | date:'mm'}}         | ‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM
"2016-06-21T10:47:40Z"    | {{mydate | date:'ss'}}            | ‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM

"2016-06-21T10:47:40Z"    | {{mydate | date:'HH:mm'}}   | ‎12‎:‎00:‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM 
"2016-06-21T10:47:40Z"    | {{mydate | date:'mm:ss'}}    | ‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM:‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM
"2016-06-21T10:47:40Z"    | {{mydate | date:'HH:ss'}}     | ‎12‎:‎00:‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM

"2016-06-21T10:47:40Z"    | {{mydate | date:'HH:mm:ss'}}   | ‎‎12‎:‎00:‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM:‎6‎/‎21‎/‎2016‎ ‎12‎:‎47‎:‎40‎ ‎PM

It seems that the minutes and seconds produce an entire date instead of the desired minutes/seconds.

**Expected/desired behavior**
The expected behaviour is that mm only returns the minutes and ss only returns the seconds.

**Reproduction of the problem**
http://plnkr.co/edit/kc8ZEMAvsx1tSbsbXTuv?p=preview

**What is the expected behavior?**
See above.

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**
- **Angular version:** 2.0.0-rc.2 && 2.0.0-rc.3
- **Browser:** [ IE 11.0.9600.18350 && Edge ] this does not occur on the latest versions of Chrome and Firefox. Haven't tested other browsers.
- **Language:** [TypeScript 1.8.10]

**Comment by laskoviymishka** [comment]
I'll take a look on this. May be there is an issue with Intl.DateTimeFormatter in IE and EDGE

**Comment by vicb** [comment]
IIRC I had to disable IE tests on our CI... IT would be good to re-enable them once you find the root cause. Thanks !

**Comment by laskoviymishka** [comment]
@vicb bug was confirme by Chakra team. So it kinda bloked.

**Comment by vicb** [comment]
> bug was confirmed by Chakra team

do you have a link ? thanks

**Comment by laskoviymishka** [comment]
@vicb _https://github.com/Microsoft/ChakraCore/issues/1223_ here it is

**Comment by felipedrumond** [comment]
Any solution guys?
If I use 
`date:"dd/MM/yy HH:mm:ss"`

It displays the time 3 times 

**05/10/16 18:00:10/5/2016 6:23:14 PM:10/5/2016 6:23:14 PM**

**Comment by jamesmcroft** [comment]
@felipedrumond I too am seeing this issue in Edge. Not sure if this is closed due to being fixed but the issue still exists.

**Comment by danjford** [comment]
Am also experiencing this in Angular 2.0.0 on IE 11:

`{{exampleDate | date: 'HH:mm'}}` if formatted as  `10:00:10/19/2016 10:00:00 AM`.

**Comment by felipedrumond** [comment]
I had to provide a solution by creating my own pipe. It is simple and it solves my problem while there is no fix for this issue:

`import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
    name: 'customDate'
})

export class DataPipe implements PipeTransform {
    transform(value: string, type: string = ""): string {

        let dd = value.substr(8, 2);
        let MM = value.substr(5, 2);
        let yyyy = value.substr(0, 4);
        let date = `${dd}/${MM}/${yyyy}`;

        let time = '';

        if (type != 'short') {
            let hh = value.substr(11, 2);
            let mm = value.substr(14, 2);
            let ss = value.substr(17, 2);
            time = `${hh}:${mm}:${ss}`;
        }

        return `${date} ${time}`;
    }
}`

This pipe displays dd/MM/yyyy or dd/MM/yyyy HH:mm:ss

**Comment by jamesmcroft** [comment]
There is also another work around for this folks which uses the Date pipe as it is something we had to do in our project (it's not perfect but it's better than being broken)

I recommend using the 'short' (9/3/2010, 12:05 PM) or 'shortDate' (9/3/2010) or 'shortTime' (12:05 PM). These do seem to work in IE and Edge, as well as other browsers. They also support locale so you'll get a formatted date/time for that area. It's a good workaround while this gets fixed.

**Comment by mrsheepuk** [comment]
Another workaround, if you have moment available, the following creates a datex pipe which uses the moment formatter instead of the angular one:

```
import { Pipe, PipeTransform } from '@angular/core';
import * as moment from 'moment';

@Pipe({
    name: 'datex'
})

export class DatexPipe implements PipeTransform {
    transform(value: Date, format: string = ""): string {
        return moment(value).format(format);
    }
}
```

Usage:

`{{exampleDate | datex:'DD/MM/YYYY HH:mm:ss'}}`

This will only work correctly where exampleDate is something which can be parsed by the default moment parsing rules - see [the moment documentation](http://momentjs.com/docs/#/parsing/string/).

**Comment by gary-b** [comment]
So when the user provides a custom format string like 'mm:ss' angular's Intl.ts file will call Intl.DateTimeFormat.format() multiple times to extract each time component individually. (As an aside, [the proposed ECMAScript 4 formatToParts method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat/formatToParts) would probably eliminate the need for these calls.)

Along with minute and seconds issue described here, it also seems you always get the minutes component back when requesting the 24 format hour in Edge/IE.

I've got a fix locally for these issues, using Date's get{TimeComponent}() functions in the broken scenarios.

@vicb Would an approach like this be suitable for PR?

**Comment by pkozlowski-opensource** [comment]
> I've got a fix locally for these issues, using Date's get{TimeComponent}() functions in the broken scenarios.

@gary-b worth trying although this would mean that we need to "hard-code" / implement padding and time separators, right? I don't know enough about possible formats and display across all the locales of the World to know if we can do this "safely"....

**Comment by flavioafc** [comment]
This pipe uses the Internationalization API. Therefore it is only reliable in Chrome and Opera browsers.
The solution of @felipedrumond works well.

**Comment by ckuetbach** [comment]
#11054

**Comment by liorwohl** [comment]
Still present in 4.0.3 release..

**Comment by MISRV001** [comment]
Is there any ETA to fix this issue?

**Comment by mischkl** [comment]
Hello, what's happening with this and how are people dealing with it? Has everyone just decided to ignore Angular's built-in date pipe since it doesn't work properly?

**Comment by mrsheepuk** [comment]
We're still using my moment-based workaround listed further up in the comments for this issue (see https://stackoverflow.com/a/40160501/789529 for the latest version thereof).

Unless you can be certain none of your users are using IE, the built-in formatter is essentially useless while this issue is unaddressed.

**Comment by btaens** [comment]
Still very much broken in 4.2.0-rc.2.

**Comment by JamshedLatipov** [comment]
In mobile firefox v53.0.2 does not work to.

**Comment by robert197** [comment]
...waiting until this branch will be merged and released :+1:

**Comment by MickL** [comment]
The pull request contains breaking changes. Hope it will be 5.0 then, not 6.0!

**Comment by MickL** [comment]
Looks awesome. Hoping for a blog post on angular.io :)

**Comment by dalu** [comment]
@ocombe where should I look for said blog post?

**Comment by ocombe** [comment]
https://blog.angular.io/version-5-0-0-of-angular-now-available-37e414935ced
